docs/source-1.0/spec/core/behavior-traits.rst

```diff
@@ -521,3 +521,195 @@ See
+    that applied the encodings MUST generate a Content-Encoding header field
+    that lists the content codings in the order in which they were applied.
+
+If a request has a structure member bound to the ``Content-Encoding`` header,
+the trait MAY still be applied. If the request is compressed, then the
+compression encoding MUST be appended to the ``Content-Encoding`` header after
+any user-provided encoding(s).
+
+As an example, using the ``PutWithContentEncoding`` operation below with the
+``customEncoding`` member set to ``brotli``, the header would be
+``"Content-Encoding": "brotli, gzip"``.
+
+.. code-block:: smithy
+
+    namespace smithy.example
+
+    @requestCompression(
+        encodings: ["gzip"]
+    )
+    operation PutWithContentEncoding {
+        input: PutWithContentEncodingInput
+    }
+
+    @input
+    structure PutWithContentEncodingInput {
+        @httpHeader("Content-Encoding")
+        customEncoding: String, // brotli
+
+        @httpPayload
+        data: String
+    }
+
+Clients MUST encode the request content prior to request signing, otherwise the
+request signature will mismatch.
+
+Because small payloads may become longer when compressed, operations SHOULD only
+be compressed when the payload size is greater than or equal to the value of the
+minimum compression threshold (see ``REQUEST_MIN_COMPRESSION_SIZE_BYTES`` in
+:ref:`compression-client-implementation`).
+
+If the request contains a member with the :ref:`streaming-trait` applied without
+the :ref:`requiresLength-trait` applied, the request MUST be compressed if
+enabled regardless of the minimum compression threshold.
+
+Operations with input members with both :ref:`streaming-trait` and
+:ref:`requiresLength-trait` are not allowed to use request compression. This
+avoids the client reading and compressing the entire stream for the length of
+the compressed data to set the ``Content-Length`` header.
+
+
+.. _compression-client-implementation:
+
+Client Implementation
+=====================
+
+.. warning::
+
+    The client configuration settings defined in this specification are still
+    evolving and subject to change.
+
+Smithy clients MUST expose a setting ``DISABLE_REQUEST_COMPRESSION`` to disable
+request compression. The value MUST be a boolean value which defaults to
+``false``. In order of precedence from highest to lowest, the setting shall be
+exposed:
+
+- Per-request configuration (SHOULD, if request-level configuration is supported
+  by the client)
+- Service client configuration - ``DisableRequestCompression`` (MUST, with
+  idiomatic casing/spacing for that client programming language)
+
+Smithy clients MUST expose a setting ``REQUEST_MIN_COMPRESSION_SIZE_BYTES`` to
+specify the minimum size in bytes that a request body should be to trigger
+compression. The value MUST be a non-negative integer in the range ``0`` and
+``10485760`` inclusive, and defaults to ``10240`` bytes. If the value is outside
```

Comment:
I personally feel that adding a clarifying note about why we choose `10240` will be worth. Otherwise this feels like a magic number.